pkg/machineproviders/providers/openshift/machine/v1beta1/failuredomain/failuredomain.go

```diff
@@ -142,6 +151,109 @@ func (f failureDomain) Equal(other FailureDomain) bool {
+			return nil, fmt.Errorf("cannot combine failure domain with template failure domain: %w", err)
+		}
+
+		comparableFailureDomains = append(comparableFailureDomains, failureDomain)
+	}
+
+	return comparableFailureDomains, nil
+}
+
+// CombineWithTemplateFailureDomain creates a copy of templateFailureDomain and overrides any set values with the values from the current failure domain.
+func (f failureDomain) CombineWithTemplateFailureDomain(templateFailureDomain FailureDomain) (FailureDomain, error) {
+	if templateFailureDomain == nil {
+		return nil, errMissingTemplateFailureDomain
+	}
+
+	if f.platformType != templateFailureDomain.Type() {
+		return nil, errMismatchedPlatformType
+	}
+
+	switch f.platformType {
+	case configv1.AWSPlatformType:
+		return f.combineWithTemplateFailureDomainAWS(templateFailureDomain.AWS()), nil
+	case configv1.AzurePlatformType:
+		return f.combineWithTemplateFailureDomainAzure(templateFailureDomain.Azure()), nil
+	case configv1.GCPPlatformType:
+		return f.combineWithTemplateFailureDomainGCP(templateFailureDomain.GCP()), nil
+	case configv1.OpenStackPlatformType:
+		return f.combineWithTemplateFailureDomainOpenStack(templateFailureDomain.OpenStack()), nil
+	default:
+		return NewGenericFailureDomain(), nil
+	}
+}
+
+func (f failureDomain) combineWithTemplateFailureDomainAWS(templateFailureDomain machinev1.AWSFailureDomain) FailureDomain {
+	fd := templateFailureDomain.DeepCopy()
+
+	if f.aws.Placement.AvailabilityZone != "" {
+		fd.Placement = f.aws.Placement
+	}
+
+	if f.aws.Subnet != nil && !reflect.DeepEqual(f.aws.Subnet, machinev1.AWSResourceReference{}) {
+		fd.Subnet = f.aws.Subnet
+	}
+
+	return NewAWSFailureDomain(*fd)
+}
+
+func (f failureDomain) combineWithTemplateFailureDomainAzure(templateFailureDomain machinev1.AzureFailureDomain) FailureDomain {
+	fd := templateFailureDomain.DeepCopy()
+
+	if f.azure.Zone != "" {
+		fd.Zone = f.azure.Zone
+	}
+
+	if f.azure.Subnet != "" {
+		fd.Subnet = f.azure.Subnet
+	}
+
+	return NewAzureFailureDomain(*fd)
+}
+
+func (f failureDomain) combineWithTemplateFailureDomainGCP(templateFailureDomain machinev1.GCPFailureDomain) FailureDomain {
+	fd := templateFailureDomain.DeepCopy()
+
+	if f.gcp.Zone != "" {
+		fd.Zone = f.gcp.Zone
+	}
+
+	return NewGCPFailureDomain(*fd)
+}
+
+func (f failureDomain) combineWithTemplateFailureDomainOpenStack(templateFailureDomain machinev1.OpenStackFailureDomain) FailureDomain {
+	fd := templateFailureDomain.DeepCopy()
+
+	if f.openstack.AvailabilityZone != "" {
+		fd.AvailabilityZone = f.openstack.AvailabilityZone
+	}
+
+	//nolint: nestif
+	if f.openstack.RootVolume != nil {
+		if fd.RootVolume == nil {
+			fd.RootVolume = f.openstack.RootVolume
+		} else {
+			if f.openstack.RootVolume.AvailabilityZone != "" {
+				fd.RootVolume.AvailabilityZone = f.openstack.RootVolume.AvailabilityZone
+			}
+
+			if f.openstack.RootVolume.VolumeType != "" {
+				fd.RootVolume.VolumeType = f.openstack.RootVolume.VolumeType
+			}
+		}
+	}
```

Comment:
I think this is equivalent and reduces the nesting a bit, might be able to get rid of that linter marker
```suggestion
	if fd.RootVolume == nil {
		fd.RootVolume = f.openstack.RootVolume
	} else if f.openstack.RootVolume != ni {
		if f.openstack.RootVolume.AvailabilityZone != "" {
			fd.RootVolume.AvailabilityZone = f.openstack.RootVolume.AvailabilityZone
		}
	
		if f.openstack.RootVolume.VolumeType != "" {
			fd.RootVolume.VolumeType = f.openstack.RootVolume.VolumeType
		}
	}
```